src/internal.rs

```diff
@@ -349,13 +349,21 @@ impl Local {
         debug_assert_eq!(self.guard_count.get(), 0);
         debug_assert_eq!(self.handle_count.get(), 0);
 
+        // Temporarily increment handle count. This is required so that the following call to `pin`
+        // doesn't call `finalize` again.
+        self.handle_count.set(1);
+        unsafe {
+            // Pin and move the local bag into the global queue.
+            let guard = &self.pin();
+            self.global().push_bag(&mut *self.bag.get(), guard);
```

Comment:
Aha, now I can see what's wrong with the current implementation.  The problem is `push_bag()` may access the global queue's internal pointers without protection.  This patch adds that required protection.

I'd like to leave a comment that `push_bag()` will not add a garbage in the local bag so that `self.bag` is really empty after this call.

Reply:
Added a comment.